File: cross-number-algorithm/cross_number_algorithm.cpp
#include <iostream>
#include <vector>

// 点の座標クラス
class Point {
 public:
  int32_t x;
  int32_t y;

  explicit Point(const int32_t x, const int32_t y) : x(input_x), y(input_y) {}
}

// 多角形クラス
class Polygon {
 public:
  std::vector<Point> vertexes;

  explicit Polygon(const std::vector<Point>& points) : vertexes(points) {}
}

/**
 * 点が多角形内にあるか判定する関数
 * 
 * @return true(内側) / false(外側)
 */
bool is_point_inside_polygon(const Point& point, const Polygon& polygon) {
  // 交差回数のカウンタ
  uint32_t c = 0;

  for (const auto& v : polygon.vetexes) {
    const uint32_t i = &v - &polygon.vetexes[0];

    // 隣の頂点
    const Point v1 = (i == polygon.vetexes.size() - 1) ? polygon.vetexes[0]
                                                       : polygon.vetexes[i + 1];
    // pointのy座標が2頂点の間にあるか？
    if (((v.y <= point.y) && (v1.y > point.y)) ||
        ((v.y > point.y) && (v1.y <= point.y))) {
      double x = v.x + ((point.y - v.y) / double(v1.y - v.y) * (v1.x - v.x));
      
      // 辺上にある場合は内側判定
      if(point.x == x){
        return true;
      }

      if (point.x < x) {
        count++;
      }
    }
  }

  return (count % 2) == 1;
}

// 使用例
int main() {
  std::vector<Point> ps = {{1, 1}, {4, 1}, {4, 4}, {1, 4}};
  Polygon polygon{ps};

  Point p1{2, 2};

  if (is_point_inside_polygon(p1, polygon)) {
    std::cout << "inside!" << std::endl;
  } else {
    std::cout << "outside!" << std::endl;
  }

  return 0;
}

File: cross_number_algorithm/cross_number_algorithm.py
def is_point_in_polygon(point, polygon):
    # 交差回数のカウンタ
    count = 0

    p_x = point[0]
    p_y = point[1]

    for i, v, in enumerate(polygon):
        # 頂点座標
        v_x = v[0]
        v_y = v[1]
        # 隣の頂点座標
        v1_x = polygon[0][0] if i == len(polygon)-1 else polygon[i+1][0]
        v1_y = polygon[0][1] if i == len(polygon)-1 else polygon[i+1][1]

        if((v_y <= p_y) and (v1_y > p_y)) or ((v_y > p_y) and (v1_y <= p_y)):
            x = v_x + ((p_y - v_y) / (v1_y - v_y)*(v1_x -v_x))
            # 辺上にある場合は内側判定
            if x == p_x:
                return True
            elif p_x < x:
                count += 1

    return bool(count%2 == 1)

# 使用例
if __name__ == '__main__':
    point = [2,2]
    polygon = [[1, 1], [4, 1], [4, 4], [1, 4]]

    if is_point_in_polygon(point, polygon):
        print("inside!")
    else:
        print("outside")
